### PR TITLE
Extend re-entrancy detection for broken trace chains

### PR DESCRIPTION
**Pull request by gnodet**
## Summary

The trace-based re-entrancy detection added in PR #1957 does not fully cover all code paths. Specifically, Maven 4's `RequestTraceHelper` converts between Maven API traces and resolver `RequestTrace` objects via `toResolver()`, creating a fresh trace chain that loses the `REPOSITORY_SYSTEM_CALL` marker. This causes `MavenValidator` to reject the collect request in the `MavenITgh12305` integration test (PR apache/maven#12481) with `Invalid Collect Request: null -> []`.

This PR addresses the gap with two complementary fixes:

- **Session-scoped re-entrancy detection** (`DefaultRepositorySystem`): Each public method now maintains a depth counter in `SessionData` via `enterSessionScope()`/try-finally decrement. When the counter is > 0 on entry, the call is detected as re-entrant regardless of whether the `RequestTrace` chain carries the marker. This catches cases where consumers rebuild the trace chain from a different tracing system.

- **Skip managed dependency validation** (`DefaultRepositorySystemValidator`): Managed dependencies are declarative constraints that only take effect when a matching dependency is encountered during collection. Validating them eagerly rejects valid builds where a BOM imports managed deps with uninterpolated expressions (e.g. `${osgi.version}`) that are never actually used. If a managed dependency IS matched and its coordinates are invalid, the error surfaces naturally during version/artifact resolution.

Together these changes allow [apache/maven#12481](https://github.com/apache/maven/pull/12481) to remove the consumer-side workarounds for uninterpolated expressions.

## Context

- [PR #1957](https://github.com/apache/maven-resolver/pull/1957) — original re-entrancy detection (trace-based)
- [apache/maven#12481](https://github.com/apache/maven/pull/12481) — Maven PR that removes consumer-side workarounds
- [MavenITgh12305](https://github.com/apache/maven/blob/master/its/core-it-suite/src/test/java/org/apache/maven/it/MavenITgh12305InvalidCollectRequestUninterpolatedManagedDepsTest.java) — the IT test that exercises this scenario

## Test plan

- [x] New test: `sessionScopedDetectionSkipsValidationWhenTraceChainIsBroken` — verifies inner calls with broken trace + uninterpolated expressions succeed via session-scoped detection
- [x] New test: `collectDependenciesAcceptsManagedDepsWithUninterpolatedExpressions` — reproduces the MavenITgh12305 scenario
- [x] New test: `collectDependenciesStillRejectsInvalidDirectDependencies` — ensures direct deps are still validated
- [x] New test: `sessionScopedDepthIsProperlyDecrementedOnExit` — verifies counter cleanup for independent calls
- [ ] CI: all existing + new tests pass

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by cstamas** [comment]
See https://github.com/apache/maven-resolver/pull/2008

MavenValidator should just not implement new method. But in some other, non-Maven use cases, uninterpolated depMgt may be seen as error.

**Comment by gnodet** [comment]
Thanks for the feedback @cstamas! You're right — removing managed dependency validation entirely is too broad.

I've reverted the managed dependency validation changes in 74005921. This PR now focuses solely on **session-scoped re-entrancy detection** (the `AtomicInteger` depth counter in session data), which addresses the broken trace chain scenario independently of the managed dependency question.

The managed dependency separation (direct vs managed validation) is better handled by #2008's `validateManagedDependency()` approach, which lets MavenValidator abstain while keeping the validation available for non-Maven consumers.

**Review by gnodet — COMMENTED** [review]
The session-scoped re-entrancy depth counter is a sound approach for catching broken trace chains. The try-finally guard pattern is consistently and correctly applied across all 8 public methods. A few observations:

**Medium severity:**
- 🧵 **Session-scoped vs thread-scoped counter**: The `AtomicInteger` depth counter in `SessionData` is shared across all threads using the same session. Re-entrancy is inherently per-call-stack, so in parallel builds with a shared session, Thread B's independent outermost call could see `depth > 0` from Thread A and skip validation ("fail-open"). Practical impact is limited since the primary trace-based `isReentrant(trace)` is per-call-stack and the session counter is a fallback for the Maven 4 `RequestTraceHelper` edge case (single-threaded model builder). A `ThreadLocal` would be more semantically correct but may be over-engineering for this specific use case.

**Low severity:**
- `fail("msg: " + e)` at test line 340 discards the exception stack trace. JUnit 5's `fail(String, Throwable)` overload preserves the full context for easier diagnosis.
- No test verifies that the depth counter is properly decremented when the delegate throws an exception. The try-finally pattern guarantees correctness, but a test proving cleanup on error would strengthen confidence.

**Strengths:**
- Clean revert of the managed dependency validation concern to the already-merged PR #2008.
- Consistent try-finally exitGuard across all 8 relevant methods; terminal operations (`install`, `deploy`, etc.) correctly left unmodified.
- The anonymous `Object` subclass with `toString()` for the session data key is a good pattern for identity-based key safety.

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of Guillaume Nodet_

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultRepositorySystem.java [hunk 1]
```diff
@@ -587,6 +670,41 @@ private static boolean isReentrant(RequestTrace trace) {
...
+     */
+    private static Runnable enterSessionScope(RepositorySystemSession session) {
+        AtomicInteger depth = getReentryDepth(session);
+        depth.incrementAndGet();
+        return depth::decrementAndGet;
+    }
+
+    private static AtomicInteger getReentryDepth(RepositorySystemSession session) {
+        return (AtomicInteger) session.getData().computeIfAbsent(SESSION_REENTRY_DEPTH_KEY, () -> new AtomicInteger(0));
+    }
```
[the remark this answers is not in the document]
Comment (gnodet) [reply maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultRepositorySystem.java:697]: Fixed in 48c7a292. Replaced the session-scoped `AtomicInteger` with a `ThreadLocal<int[]>` — re-entrancy is per-call-stack, so thread-scoped is the correct semantic. This avoids false positives in parallel builds where multiple threads share a session.

File: maven-resolver-impl/src/test/java/org/eclipse/aether/internal/impl/DefaultRepositorySystemReentrancyTest.java [hunk 2]
```diff
@@ -303,4 +312,120 @@ void outerCallWithNullTraceStillStampsMarker() throws Exception {
...
+            // expression. If session-scoped detection fails, the validator rejects this.
+            VersionRangeRequest innerRequest = new VersionRangeRequest(
+                    new DefaultArtifact("g:inner:${project.version}"), Collections.emptyList(), null);
+            // Fresh trace — no REPOSITORY_SYSTEM_CALL marker (simulates Maven's trace conversion)
+            innerRequest.setTrace(RequestTrace.newChild(null, "MavenModelResolver"));
+            try {
+                systemRef.get().resolveVersionRange(s, innerRequest);
+            } catch (Exception e) {
+                fail("Re-entrant call with broken trace chain should succeed via session-scoped detection: " + e);
+            }
```
[the remark this answers is not in the document]
Comment (gnodet) [reply maven-resolver-impl/src/test/java/org/eclipse/aether/internal/impl/DefaultRepositorySystemReentrancyTest.java:340]: Fixed in 48c7a292. Changed to `fail(msg, e)` to preserve the full stack trace.

**Review by gnodet — COMMENTED** [review]
**Re-review** after latest commit (48c7a29).

All three concerns from the previous review have been directly and correctly addressed:

1. **Thread-scoped depth counter** — session-scoped `AtomicInteger` replaced with `ThreadLocal<int[]>`, giving per-call-stack semantics that avoid false positives in parallel builds. The `int[]` pattern avoids boxing overhead and makes the captured-`Runnable` decrement trivial.

2. **Stack trace preservation** — `fail(msg + e)` changed to `fail(msg, e)` to preserve exception stack traces.

3. **Exception cleanup test** — new test `threadScopedDepthIsDecrementedWhenDelegateThrows` confirms the try-finally guard works on the exception path.

The `Runnable`-based `exitGuard` with try-finally is consistently applied across all 8 relevant public methods. Terminal operations (install, deploy, etc.) are correctly left unmodified.

A couple of minor, non-blocking observations:

- The method `enterSessionScope` no longer touches session state — it operates on a `ThreadLocal`. The name is a holdover from the previous implementation. Consider renaming to `enterReentrantScope` or `enterThreadScope` to match the actual semantics.
- The PR branch is behind master due to the #2008 merge (validator factory abstain support). A rebase will be needed before merge, with mechanical updates to the test constructors (`singletonList` → `singletonMap`).
- The PR description still mentions "Skip managed dependency validation" as a change, but that was reverted in commit 74005921. The description should be updated to reflect the current scope (thread-scoped re-entrancy detection only).

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of gnodet_

**Review by gnodet — COMMENTED** [review]
**Re-review** (3rd pass) after commit 94d9fda5.

The only change since the last review is a trivial spotless formatting fix (line wrapping in one test variable declaration). The full implementation remains correct:

- `ThreadLocal`-based per-thread depth tracking avoids false positives in parallel builds
- Consistent try-finally guards across all 8 re-entrancy-aware public methods
- Comprehensive test coverage including exception safety path
- CI is fully green across all matrix jobs

All concerns from previous reviews remain addressed. Rebase against master is still needed for the #2008 constructor signature change (`List<ValidatorFactory>` → `Map<String, ValidatorFactory>`) but the update is mechanical.

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of gnodet_

**Review by cstamas — APPROVED** [review]

**Review by gnodet — COMMENTED** [review]
The ThreadLocal-based re-entrancy detection for broken trace chains looks correct and well-tested. The evolution from session-scoped AtomicInteger to ThreadLocal<int[]> properly addresses the concurrency concern. A few minor observations:

- The `session` parameter in `isReentrant(RequestTrace, RepositorySystemSession)` is unused (private method, no API contract to maintain)
- `enterSessionScope`/`exitSessionScope` naming is slightly misleading since the method now operates on a ThreadLocal, not session state
- The static `ThreadLocal<int[]> REENTRY_DEPTH` is never cleaned up via `remove()` when depth reaches 0 — negligible memory impact but calling `remove()` in the finally block would be more hygienic for thread-pooled environments

_This review was generated by an AI agent and may contain inaccuracies. Please verify all suggestions before applying._

_Claude Code on behalf of gnodet_
